bittide/src/Bittide/ClockControl/Si539xSpi.hs

```diff
@@ -0,0 +1,280 @@
+        | b == byte -> FetchReg (succ i)
+        | otherwise -> Error i
+
+      (FetchReg i, _)                   -> WriteEntry i
+      (Wait ((==maxBound) -> True) i, _)-> FetchReg (succ i)
+      (Wait j i, _)                     -> Wait (succ j) i
+      (WriteEntry _, Nothing)           -> currentState
+      (ReadEntry _, Nothing)            -> currentState
+      (Finished , _)                    -> currentState
+      (Error _ , _)                     -> currentState
+
+    spiOp = case currentState of
+      FetchReg _   -> Nothing
+      WriteEntry _ -> Just RegisterOperation{regPage, regAddress, regWrite = Just byte}
+      ReadEntry _  -> Just RegisterOperation{regPage, regAddress, regWrite = Nothing}
+      Wait _ _     -> Nothing
+      Finished     -> extSpi
+      Error _      -> extSpi
+
+    (busy, returnedByte) = case currentState of
+      Finished -> (spiBusy, spiByte)
+      Error _  -> (spiBusy, spiByte)
+      _        -> (True, Nothing)
+
+-- | Keeps track of the current 'Page' and 'Address' of the @Si539x@ chip.
+-- Furthermore is used to track the current communication cycle.
+data DriverState dom = DriverState
+  { currentPage         :: Maybe Page
+  -- ^ Current 'Page' of the @Si539x@ chip.
+  , currentAddress      :: Maybe Address
+  -- ^ Current 'Address' of the @Si539x@ chip.
+  , currentOp           :: Maybe RegisterOperation
+  -- ^ Current communication transaction.
+  , commandAcknowledged :: Acknowledge
+  -- ^ Whether or not the current transaction has already been acknowledged.
+  , idleCycles          :: Index (PeriodCycles dom (Nanoseconds 95))
+  -- ^ After communication, slave select must be high for at least 95ns.
+  }
+  deriving (Generic, NFDataX)
+
+-- | Circuitry that controls an spi core based on a state machine that ensures communication
+-- transactions with an @Si539x@ chip are executed correctly. It makes sure communication
+-- operations target the right register and communication operations are spaced correctly.
+si539xSpiDriver ::
+  forall dom minTargetPeriodPs .
+  (HiddenClockResetEnable dom, KnownNat (DomainPeriod dom)) =>
+  -- | Minimum period of the spi clock frequency for the spi clock divider.
+  SNat minTargetPeriodPs ->
+  -- | Read or write operation for the @Si539X@ registers.
+  Signal dom (Maybe RegisterOperation) ->
+  -- | MISO
+  "MISO" ::: Signal dom Bit ->
+  -- |
+  -- 1. Byte returned by read / write operation.
+  -- 2. The spi interface is 'Busy' and does not accept new operations.
+  -- 3. Outgoing SPI signals: (SCK, MOSI, SS)
+  ( Signal dom (Maybe Byte)
+  , Signal dom Busy
+  , ( "SCK"  ::: Signal dom Bool
+    , "MOSI" ::: Signal dom Bit
+    , "SS"   ::: Signal dom Bool
+    )
+  )
+si539xSpiDriver SNat incomingOpS miso = (fromSlave, decoderBusy, spiOut)
+ where
+  spiOut = (sck, mosi, ss)
+  (sck, mosi, ss, spiBusyS, acknowledge, receivedData) =
+    spiMaster SPIMode0 (SNat @(HalfPeriodCycles dom minTargetPeriodPs)) d1 spiWrite miso
```

Comment:
Could you add a comment explaining why you pass _half_ this?

`HalfPeriodCycles` feel like a weird definition to make, is this a common thing to do? Wouldn't `DivRU (PeriodToCycles dom minTargetPeriodPs) 2` be clearer?